cli/flox/src/utils/message.rs

```diff
@@ -47,3 +50,173 @@ pub(crate) fn package_installed(pkg: &PackageToInstall, environment_description:
         pkg.id()
     ));
 }
+
+/// A test spy for messages printed to stderr through the `message` module.
+/// In unit tests, the messaging functions of the `message` module will,
+/// populate a `History` in addition to printing the message to stderr.
+/// This allows the test to assert against the messages printed to stderr,
+/// without refactoring existing implementations to print to a `Write` trait
+/// and injecting a mock in tests.
+///
+/// The `History` is thus comparitivly non-intrusive.
+/// However, it comes with a few caveats:
+///
+/// Messaging functions are currently effectively (process) globals,
+/// inheriting the global nature of `stderr`.
+/// `#[test]` functions are run in parallel on separate _threads_
+/// by both `cargo test` and `cargo nextest`.
+/// Assuming an equally global `History`, messages from one test
+/// will be visible in another test.
+/// There are several approaches to avoid this:
+///
+/// * running tests in sequence (e.g. `cargo test -- --test-threads=1`),
+///   slows down the test suite.
+/// * additional bookkeeping of scopes (e.g. via `tracing`),
+///   complicates the test setup, wither needs to be done manually,
+///   or we maintain a `#[test]` replacement. (That said,
+///   there may be something to learn from `tracing`s architecture).
+/// * Pass a `History` instance to the messaging functions,
+///   intrusive as it reqeuries all callers to be updated.
+/// * Use a [thread_local!] `History`, which is the current approach.
+///
+/// Being thread local, the `History` is not shared between threads,
+/// i.e. not shared between tests.
+/// On the flip side, if the test spawns additional threads,
+/// which print messages, these will be captured in the `History`
+/// of the thread that spawned them, rather than the test's `History`.
+/// Likewise if the test uses a multi-threaded async runtime, like `tokio`,
+/// messages from spawned tasks will not be captured in the `History`.
+/// Also due to runtimes sharing threads between tasks, the `History` available
+/// in the task is also shared between tasks.
+///
+/// Luckily, our CLI surface is largely single threaded.
+/// Threads are mainly spawned for blocking operations either in the library layer,
+/// or to run library funcitons in the background, e.g. to print a progress bar
+/// on the main thread.
+/// Also, while technically within an async runtime, the CLI is largely synchronous,
+/// and the default [#[tokio::test]] runtime is single threaded.
+#[cfg(test)]
+pub mod history {
+    use std::cell::RefCell;
+    use std::collections::VecDeque;
+
+    thread_local! {
+        static THREAD_HISTORY: RefCell<History> = const {
+            RefCell::new(History {
+                messages:VecDeque::new(),
+            })
+        };
+    }
+
+    pub(crate) struct History {
+        messages: VecDeque<String>,
```

Comment:
Are you expecting to remove messages from the front instead of iterating over them?